src/cs/Ssh/SecureStream.cs

```diff
@@ -0,0 +1,291 @@
+		{
+			SshConnectionException? ex = null;
+			var clientSession = (SshClientSession)this.session;
+			if (!(await clientSession.AuthenticateServerAsync(cancellation).ConfigureAwait(false)))
+			{
+				ex = new SshConnectionException(
+					"Server authentication failed.", SshDisconnectReason.HostKeyNotVerifiable);
+			}
+
+			if (ex == null && !(await clientSession.AuthenticateClientAsync(
+				this.clientCredentials, cancellation).ConfigureAwait(false)))
+			{
+				ex = new SshConnectionException(
+					"Client authentication failed.", SshDisconnectReason.NoMoreAuthMethodsAvailable);
+			}
+
+			if (ex != null)
+			{
+				await this.session.CloseAsync(ex.DisconnectReason, ex).ConfigureAwait(false);
+				throw ex;
+			}
+
+			channel = await this.session.OpenChannelAsync(cancellation).ConfigureAwait(false);
+		}
+		else
+		{
+			channel = await this.session.AcceptChannelAsync(cancellation).ConfigureAwait(false);
+		}
+
+		this.stream = CreateStream(channel);
+		channel.Closed += (_, _) =>
+		{
+			this.Dispose();
+		};
+	}
+
+	/// <summary>
+	/// Creates a stream instance for a channel. May be overridden to create a
+	/// <see cref="SshStream" /> subclass.
+	/// </summary>
+	protected virtual SshStream CreateStream(SshChannel channel)
+	{
+		return new SshStream(channel);
+	}
+
+	protected override void Dispose(bool disposing)
+	{
+		if (disposing)
+		{
+			this.stream?.Dispose();
+			this.session.Dispose();
+
+			// If the session has not connected yet, it doesn't know about the stream and won't dispose it.
+			// So we dispose it explicitly here.
+			this.transportStream.Dispose();
+		}
+
+		base.Dispose(disposing);
+	}
+
+	/// <summary>
+	/// Close the SSH session with <see cref="SshDisconnectReason.None"/> reason and dispose the underlying transport stream.
+	/// </summary>
+	public async Task CloseAsync()
+	{
+		await this.session.CloseAsync(SshDisconnectReason.None, this.session.GetType().Name + " disposed").ConfigureAwait(false);
+		this.session.Dispose();
+
+#if !NETSTANDARD2_0
+		await this.transportStream.DisposeAsync().ConfigureAwait(false);
+#else
+		this.transportStream.Dispose();
+#endif
+	}
+
+	/// <summary>
+	/// The SSH software name and version of the remote client,
+	/// parsed when a connection is made to the server
+	/// </summary>
+	public SshVersionInfo? RemoteVersion => this.session.RemoteVersion;
+
+	private void OnSessionAuthenticating(object? sender, SshAuthenticatingEventArgs e)
+	{
+		Authenticating?.Invoke(this, e);
+	}
+
+	private void OnSessionClosed(object? sender, SshSessionClosedEventArgs e)
+	{
+		this.session.Closed -= OnSessionClosed;
+		Closed?.Invoke(this, e);
+	}
+
+	public override bool CanRead => true;
+
+	public override bool CanWrite => true;
+
+	public override void Flush()
+	{
+		this.stream?.Flush();
+	}
+
+	public override int Read(byte[] buffer, int offset, int count)
+	{
+		var stream = this.stream ?? throw new InvalidOperationException("Stream is not connected.");
```

Comment:
This can be encapsulated into a property

Reply:
Done.